lib/internal/process/finalization.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+// This file is a modified version of the on-exit-leak-free module on npm.
+
+const {
+  ArrayPrototypeFilter,
+  ArrayPrototypeIndexOf,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  SafeFinalizationRegistry,
+  WeakRef,
+} = primordials;
+const { validateObject, kValidateObjectAllowFunction } = require('internal/validators');
+
+function createFinalization() {
+  /**
+   * @type {SafeFinalizationRegistry}
+   */
+  let registry = null;
+
+  const refs = {
+    exit: [],
+    beforeExit: [],
+  };
+
+  const functions = {
+    exit: onExit,
+    beforeExit: onBeforeExit,
+  };
+
+  function install(event) {
+    if (refs[event].length > 0) {
+      return;
+    }
+
+    process.on(event, functions[event]);
+  }
+
+  function uninstall(event) {
+    if (refs[event].length > 0) {
+      return;
+    }
+
+    process.removeListener(event, functions[event]);
+
+    if (refs.exit.length === 0 && refs.beforeExit.length === 0) {
+      registry = null;
+    }
+  }
+
+  function ensureRegistry() {
+    if (!registry) {
+      registry = new SafeFinalizationRegistry(clear);
+    }
```

Comment:
nit (I don't know if it is any better, but I prefer if due less nesting)
```suggestion
    registry ||= new SafeFinalizationRegistry(clear);
```